### PR TITLE
Flight mode object next

### PR DESCRIPTION
**Pull request by peterbarker**
This is https://github.com/ArduPilot/ardupilot/pull/3802 and some follow-on-work which shows incremental cleanup of the code post FlightMode object creation.

### CODE REVIEW

File: ArduCopter/FlightMode.h [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        G_Dt(_copter.G_Dt),
+        ap(_copter.ap),
+        takeoff_state(_copter.takeoff_state),
+        ekfGndSpdLimit(_copter.ekfGndSpdLimit),
+        ekfNavVelGainScaler(_copter.ekfNavVelGainScaler)
+        { };
+
+protected:
+
+    virtual bool init(bool ignore_checks) = 0; // should be called at 100hz or more
```
Comment (khancyr) [inline ArduCopter/FlightMode.h]: I doubt we should call init at 100Hz !

File: ArduCopter/FlightMode.h [hunk 2]
```diff
@@ -57,7 +56,7 @@ class FlightMode {
     AC_WPNav *&wp_nav;
     AC_PosControl *&pos_control;
     AP_InertialNav &inertial_nav;
-    AP_AHRS &ahrs;
+    AP_AHRS_NavEKF &ahrs;
```
Comment (khancyr) [inline ArduCopter/FlightMode.h]: humm are you sure about that ? it seems a big change

File: ArduCopter/FlightMode.h (hunk 1 above)
Comment (peterbarker) [inline ArduCopter/FlightMode.h]: Urp.  Missed!  Thanks.

File: ArduCopter/FlightMode.h (hunk 2 above)
Comment (peterbarker) [inline ArduCopter/FlightMode.h]: Yes, but it actually does get passed an AHRS_EKF in Copter.

This change was required as part of moving the call which fills in the optical flow scalars into here.

I've already tried to point out that that call shouldn't be made how it is ATM; while cheap in the non-OF case, it should be called at all if we're never going to use the values (PosControl seems to run at 50Hz, we calculate at 400Hz, and e.g. ACRO will never use the values!)

**Comment by amilcarlucas** [comment]
Any plans for merging this one?
It's painfull to get fence worhing for all flight modes with the current code.
This PR should help a lot with that.

**Comment by OXINARF** [comment]
@khancyr You can close things 🙂 

This work has been merged to master in following PRs, closing.

**Comment by peterbarker** [comment]
On Thu, 11 Jan 2018, Francisco Ferreira wrote:

> Closed #5839.

Actually, I think the "move EKF scalars" thing isn't in yet, but that's 
cool.

**Pull request closed without merging**